### PR TITLE
feat: insforge-gen-types — TypeScript codegen from live schema

### LINKED ISSUE #1263: [Feature]: `insforge gen types typescript` — codegen TS types from live schema (Supabase parity)
## TL;DR

Add an `insforge gen types typescript` CLI command that introspects the local (or remote) Postgres schema and emits a TypeScript module typing every user table, view, RPC, and enum — Supabase's `supabase gen types typescript` equivalent. This is the one daily-DX moment from Supabase that InsForge currently lacks; the introspection plumbing already exists, only the formatter is missing.

Tracked as item **1.4** under the broader CLI parity umbrella in #1129 — filing a dedicated issue so this can ship independently.

## Problem

Today an InsForge SDK user writes types for their own tables by hand. The dashboard's schema and the client's `Database` type drift the moment anyone adds a column without remembering to update both — a class of bug that disappears the second types are generated from the live schema. Every competing platform (Supabase, PlanetScale, Convex, Neon) ships this; the absence reads as "early-stage tooling" at evaluation time.

InsForge already has the hard part: `databaseService.getSchemas()` in `backend/src/services/database/database.service.ts:28` plus the `information_schema.columns` query at `backend/src/services/database/database-advance.service.ts:186` cover the metadata needed. What's missing is a formatter on top and a CLI entry point.

### Command shape

```bash
insforge gen types typescript --local --schema public > types/database.ts
insforge gen types typescript --linked --schema public,analytics > types/database.ts
insforge gen types typescript --project-id <ref> --schema public > types/database.ts
```

Match Supabase's flags so muscle memory transfers:
- `--local` — introspect the local Docker stack
- `--linked` — use the project linked via `insforge link`
- `--project-id <ref>` — explicit remote project
- `--schema <list>` — comma-separated schemas (default `public`)
- `--postgres-url <url>` — bypass auth, point at any Postgres directly
- `-o, --output <file>` — write to file instead of stdout

### Output shape

A single `Database` interface that the SDK's typed client picks up via generics, e.g.:

```ts
export type Json = string | number | boolean | null | { [k: string]: Json } | Json[];

export interface Database {
  public: {
    Tables: {
      posts: {
        Row:    { id: string; title: string; author_id: string | null; created_at: string };
        Insert: { id?: string; title: string; author_id?: string | null; created_at?: string };
        Update: { id?: string; title?: string; author_id?: string | null; created_at?: string };
        Relationships: [{ foreignKeyName: "posts_author_id_fkey"; columns: ["author_id"]; referencedRelation: "users"; referencedColumns: ["id"]; }];
      };
    };
    Views:     { /* ... */ };
    Functions: { /* RPCs with Args + Returns */ };
    Enums:     { /* enum unions */ };
    CompositeTypes: { /* ... */ };
  };
}
```

Then `createClient<Database>(...)` gives autocomplete on `.from('posts').select(...)` everywhere.

### Implementation sketch

1. Reuse the existing introspection in `database.service.ts` / `database-advance.service.ts`. Add columns we don't already select: `is_nullable`, `column_default`, `is_identity`, `is_generated`, `udt_name`, plus FK metadata from `information_schema.key_column_usage` / `referential_constraints`, enum members from `pg_type` + `pg_enum`, function signatures from `pg_proc`.
2. Map PG types → TS:
   - `int2/4/8`, `numeric`, `real`, `double precision` → `number`
   - `int8`/`numeric` with precision > JS safe range → emit `number` but document caveat
   - `text`, `varchar`, `uuid`, `*timestamp*`, `date`, `time` → `string`
   - `bool` → `boolean`
   - `json`/`jsonb` → `Json`
   - `<enum>` → string union from `pg_enum`
   - arrays → `T[]`
   - composite types → nested interface
   - unknown → `unknown` with a `// TODO` marker
3. Nullable column → `| null` in `Row`. Default-having column → optional in `Insert`. All columns optional in `Update`.
4. Format with a deterministic alphabetical sort + Prettier so re-runs produce zero-diff output when the schema hasn't changed.
5. Ship as a sub-command of the `@insforge/cli` package (umbrella issue #1129) or, until that lands, as a standalone `@insforge/gen-types` package that the CLI later re-exports.

### Out of scope (file separate issues)

- `--lang go|swift|python|rust` — Supabase has these; one ticket per language once TS lands.
- Watch mode (`--watch`) that regenerates on migration apply.
- An MCP tool wrapper that lets agents request fresh types after a schema change.

## Why this matters

- **Closes the most-felt daily-DX gap vs Supabase.** Every "is this prod-ready?" evaluator types `supabase gen types typescript` within the first hour. The absence is conspicuous.
- **Eliminates a real bug class.** Drift between hand-authored types and live schema produces silent runtime failures the type checker should catch.
- **Cheap to ship.** Introspection is already in `database.service.ts`; this is a focused formatter + CLI wiring task. Realistically 1–2 weeks for a senior, including tests and the FK/enum/RPC edge cases.
- **Unblocks the rest of #1129's P1.** `insforge.config.toml`, `insforge link`, `db push/pull` all assume a CLI binary exists; shipping `gen types` first proves the CLI shape with a high-value command.

## Acceptance criteria

- [ ] `insforge gen types typescript --local --schema public` emits a valid TypeScript module against the docker-compose stack.
- [ ] Output covers tables, views, enums, RPCs (functions with `Args` + `Returns`), and composite types.
- [ ] `Row` / `Insert` / `Update` shapes honor nullability and defaults correctly.
- [ ] Foreign keys are present under each table's `Relationships` (mirrors Supabase shape for SDK compatibility).
- [ ] Re-running with no schema change produces byte-identical output (deterministic).
- [ ] `--linked` and `--project-id <ref>` work against a hosted InsForge project.
- [ ] `--postgres-url <url>` works against any Postgres reachable by URL.
- [ ] Docs page added under `docs/` with a copy-pasteable quickstart.
- [ ] Integration test: spin up the docker stack, apply a migration that exercises every supported PG type + an FK + an enum + an RPC, run the command, snapshot the output.

## References

- Supabase command: https://supabase.com/docs/reference/cli/supabase-gen-types
- Supabase output example: https://supabase.com/docs/guides/api/rest/generating-types
- Existing InsForge introspection: `backend/src/services/database/database.service.ts:28` (`getSchemas`) and `backend/src/services/database/database-advance.service.ts:186` (`information_schema.columns` query).
- Umbrella CLI parity issue: #1129 (this item is 1.4 in that proposal).

### PR DESCRIPTION
**Pull request by junaiddshaukat**
## Summary

New `@insforge/gen-types` package + `insforge-gen-types` CLI. Introspects a live Postgres schema and emits a deterministic TypeScript `Database` module — InsForge's equivalent of `supabase gen types typescript`.

Closes #1263.

## What is Covered

- **Tables, views, enums, functions (RPCs), composite types, FK `Relationships`**
- `Row` / `Insert` / `Update` honor nullability + defaults — identity & generated columns optional in `Insert`; non-writable views emit `Row` only
- Functions returning a table rowtype resolve to that relation's `Row`
- PG→TS mapping: numerics→`number`, `bool`→`boolean`, `json/jsonb`→`Json`, enums→string unions, arrays→`T[]`, composites→nested interfaces
- CLI flags: `--postgres-url`, `--local`, `--schema <list>`, `-o/--output`
- Deterministic output (alphabetical sort) — re-runs are byte-identical
- Docs page: `docs/core-concepts/database/type-generation.mdx`

## My Local Testing

Verified against a live Postgres 16 Docker container seeded with every supported construct (uuid/identity/enum/jsonb/array columns, FK, view, SETOF-table RPC, composite type):

| Check | Result |
|-------|--------|
| `turbo build + typecheck + lint` | ✅ 3/3 |
| Unit tests (`format.test.ts`) | ✅ 7/7 |
| Integration test vs live Postgres (`integration.test.ts`, gated by `GEN_TYPES_TEST_DB`) | ✅ 1/1 |
| Re-run determinism (`diff`) | ✅ byte-identical |
| Generated module compiles under `tsc --strict` | ✅ |
| Prettier | ✅ clean |

Run the integration test locally:
```bash
docker run -d --name gt -e POSTGRES_PASSWORD=postgres -p 5432:5432 postgres:16
GEN_TYPES_TEST_DB=postgres://postgres:postgres@localhost:5432/postgres npm test -w @insforge/gen-types
Deferred (follow-up issues)
--linked / --project-id — blocked on insforge link (#1129)
Overloaded functions collapse to one entry (last wins)
```

Cleanup docker: `docker rm -f gt-test`

image

---
## Summary by cubic
Adds `@insforge/gen-types` and the `insforge-gen-types` CLI to generate deterministic TypeScript `Database` types from a live Postgres schema, keeping client types in sync. Fulfills Linear #1263.

- **New Features**
  - Introspects tables, views, enums, functions (RPCs), and composite types; supports multi-schema.
  - Emits `Row`/`Insert`/`Update` with correct nullability and defaults; for views, `Insert` and `Update` are emitted independently based on capability.
  - Includes FK `Relationships`, preserving composite-key column order.
  - Resolves RPC return rowtypes to that relation’s `Row` within the same schema; supports `SETOF` and includes `VARIADIC`/`INOUT` input args.
  - Deterministic output (alphabetical); overloaded functions are de-duped deterministically; programmatic API `genTypes()` plus `introspect()` and `format()`.
  - CLI: `--postgres-url`, `--local`, `--schema`, `-o/--output`; validates flag values; stdout by default.
  - PG→TS mapping: numerics→`number`, `bool`→`boolean`, `json/jsonb`→`Json`, `money`→`string`, enums→string unions, arrays→`T[]`, composites→nested interfaces.
  - Docs at `docs/core-concepts/database/type-generation.mdx`; unit tests and optional live-DB integration tests.

- **Migration**
  - Install: `npm i -D @insforge/gen-types`
  - Generate: `npx insforge-gen-types --postgres-url <url> --schema public -o types/database.ts` (or `--local`)
  - Use: `import type { Database } from './types/database'` and pass `createClient<Database>(...)`

<sup>Written for commit f5e0cc1edbf274eb9a5d0d3fdad422d4912267ef. Summary will update on new commits. <a href="https://cubic.dev/pr/InsForge/InsForge/pull/1301?utm_source=github">Review in cubic</a></sup>

> [!NOTE]
> ### Add `@insforge/gen-types` package for TypeScript codegen from a live Postgres schema
> - Introduces a new `@insforge/gen-types` package with a CLI binary (`insforge-gen-types`) and programmatic API that connects to Postgres and emits a deterministic `Database` TypeScript module.
> - [`introspect.ts`](https://github.com/InsForge/InsForge/pull/1301/files#diff-b415a0adb15e78b168866cc2aa185049289315a0f32151c7f3698af2f95915a5) queries pg system catalogs to build an IR covering tables, views, enums, composite types, functions (RPCs), columns, and foreign-key relationships across one or more schemas.
> - [`format.ts`](https://github.com/InsForge/InsForge/pull/1301/files#diff-b48151bf225aaefe49fafba41276c4b809d175b3a41e01be218ac72e04299317) renders the IR into a TypeScript module with `Row`, `Insert`, and `Update` variants per table/view, enum unions, function `Args`/`Returns` types, and `Relationships` arrays, all sorted alphabetically for determinism.
> - [`typemap.ts`](https://github.com/InsForge/InsForge/pull/1301/files#diff-53d17c3150e52c782090caf3a40c8e3eeb23c28a1a258d68c7026a9d13f0ede9) maps Postgres scalar types to TypeScript primitives and resolves array, enum, and composite references to `Database[...]` paths.
> - CLI supports `--postgres-url`, `--local` (falls back to `$DATABASE_URL`), `--schema`, and `-o/--output`; documentation is added under `core-concepts/database/type-generation`.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized f5e0cc1.</sup>
> 
>

## Summary by CodeRabbit

* **New Features**
  * Added a gen-types tool and CLI to generate deterministic TypeScript Database types from Postgres schemas (tables, views, enums, functions, composite types).
* **Documentation**
  * Added a user guide and README covering installation, quickstart, CLI flags, output shape, type mappings, and wiring into SDKs; updated docs navigation.
* **Tests**
  * Added unit and Postgres-backed integration tests for formatting determinism and generated type shapes.
* **Chores**
  * Package, build, tsconfig, ignore, and task configuration for the new package.

[Review Change Stack](https://app.coderabbit.ai/change-stack/InsForge/InsForge/pull/1301?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/gen-types/src/typemap.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
+/** Postgres type-name (`udt_name` / `pg_type.typname`) -> TypeScript scalar. */
+const SCALAR_MAP: Record<string, string> = {
+  int2: 'number',
+  int4: 'number',
+  int8: 'number',
+  numeric: 'number',
+  float4: 'number',
+  float8: 'number',
+  money: 'number',
```
[the remark this answers is not in the document]
Comment (junaiddshaukat) [reply packages/gen-types/src/typemap.ts]: oh, great point fixing it

**Comment by junaiddshaukat** [comment]
@tonychang04 please have a look on this pr when you got time, thanks

File: packages/gen-types/src/cli.ts [hunk 2]
```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/env node
+import { writeFileSync } from 'node:fs';
+import { genTypes } from './index.js';
+
+const USAGE = `insforge-gen-types — generate TypeScript types from a live Postgres schema
```
Comment (tonychang04) [inline packages/gen-types/src/cli.ts:5]: should the cli be in the cli repo? maybe first look at how supabase does it

**Review by tonychang04 — CHANGES_REQUESTED** [review]
yah this feature may not be 100% needed at this moment so will postpone in a bit of reviewing
